src/util-lua-packet.c

```diff
@@ -0,0 +1,259 @@
+
+    /* proto and ports (or type/code) */
+    lua_pushinteger(luastate, p->proto);
+    if (p->proto == IPPROTO_TCP || p->proto == IPPROTO_UDP) {
+        lua_pushinteger(luastate, p->sp);
+        lua_pushinteger(luastate, p->dp);
+
+    } else if (p->proto == IPPROTO_ICMP || p->proto == IPPROTO_ICMPV6) {
+        lua_pushinteger(luastate, p->icmp_s.type);
+        lua_pushinteger(luastate, p->icmp_s.code);
+    } else {
+        lua_pushinteger(luastate, 0);
+        lua_pushinteger(luastate, 0);
+    }
+
+    return 6;
+}
+
+/** \internal
+ *  \brief get tcp/udp/sctp source port
+ *  \param luastate the lua state
+ */
+static int LuaPacketSport(lua_State *luastate)
+{
+    struct LuaPacket *s = (struct LuaPacket *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->p == NULL) {
+        LUA_ERROR("failed to get packet");
+    }
+    Packet *p = s->p;
+
+    switch (p->proto) {
+        case IPPROTO_TCP:
+        case IPPROTO_UDP:
+        case IPPROTO_SCTP:
+            lua_pushinteger(luastate, p->sp);
+            break;
+        default:
+            LUA_ERROR("sp only available for tcp, udp and sctp");
+    }
+
+    return 1;
+}
+
+/** \internal
+ *  \brief get tcp/udp/sctp dest port
+ *  \param luastate the lua state
+ */
+static int LuaPacketDport(lua_State *luastate)
+{
+    struct LuaPacket *s = (struct LuaPacket *)lua_touserdata(luastate, 1);
+    if (s == NULL || s->p == NULL) {
+        LUA_ERROR("failed to get packet");
+    }
+    Packet *p = s->p;
+
+    switch (p->proto) {
+        case IPPROTO_TCP:
+        case IPPROTO_UDP:
+        case IPPROTO_SCTP:
+            lua_pushinteger(luastate, p->dp);
+            break;
+        default:
+            LUA_ERROR("dp only available for tcp, udp and sctp");
+    }
+
+    return 1;
+}
+
+static int LuaPacketGet(lua_State *luastate)
+{
+    Packet *p = LuaStateGetPacket(luastate);
+    if (p == NULL) {
+        LUA_ERROR("failed to get packet");
+    }
+
+    struct LuaPacket *s = (struct LuaPacket *)lua_newuserdata(luastate, sizeof(*s));
+    if (s == NULL) {
+        LUA_ERROR("failed to get userdata");
+    }
+    s->p = p;
+    luaL_getmetatable(luastate, suricata_packet_metatable);
+    lua_setmetatable(luastate, -2);
+    return 1;
+}
+
+// clang-format off
+static const luaL_Reg packetlib[] = {
+    { "get", LuaPacketGet },
+    { "packet", LuaPacketPacket },
+    { "payload", LuaPacketPayload },
+    { "pcap_cnt", LuaPacketPcapCnt },
+    { "timestring", LuaPacketTimestring },
+    { "timestamp", LuaPacketTimestamp },
+    { "tuple", LuaPacketTuple },
+    { "sp", LuaPacketSport },
+    { "dp", LuaPacketDport },
+    { "__gc", LuaPacketGC },
+    { NULL, NULL }
+};
+// clang-format on
+
+int LuaLoadPacketLib(lua_State *luastate)
+{
+    luaL_newmetatable(luastate, suricata_packet_metatable);
+    lua_pushvalue(luastate, -1);
+    lua_setfield(luastate, -2, "__index");
+    luaL_setfuncs(luastate, packetlib, 0);
+    luaL_newlib(luastate, packetlib);
```

Comment:
I think the metatable functions and the lib functions should not be shared. Leads a duplicate API and confusion on how it should be used.